### PR TITLE
add-mmpose-sdk

### PR DESCRIPTION
**Pull request by irexyc**
## Motivation

Add HRNet, MSPN, LiteHRNet sdk inference support

## Modification

1. implement mmpose TopDown 2d heatmap/regression decode method.
2. implement topdown_affine transform.
3. Modify some transforms in origin mmpose pipeline. (like (ToTensor, NormalizeTensor) -> Normalize) 
4. flip_test is not supported yet. 

### update:
test results are same with pytorch for the below configs with flip_test = False
mmpose/configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/hrnet_w48_coco_256x192.py
mmpose/configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/mspn50_coco_256x192.py
mmpose/configs/body/2d_kpt_sview_rgb_img/deeppose/coco/res50_coco_256x192.py

### CODE REVIEW

File: csrc/apis/c/CMakeLists.txt [hunk 1]
```diff
@@ -24,6 +24,9 @@ else ()
         list(APPEND TASK_LIST "text_detector")
         list(APPEND TASK_LIST "text_recognizer")
     endif ()
+    if ("mmpose" IN_LIST MMDEPLOY_CODEBASES)
+        list(APPEND TASK_LIST "pose_estimator")
```
Comment (lvhan028) [inline csrc/apis/c/CMakeLists.txt:28]: Inconsistent name: `pose_detector` vs `pose_estimator`

File: csrc/codebase/mmpose/keypoints_from_heatmap.cpp [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+    int H = heatmap.shape(2);
+    int W = heatmap.shape(3);
+    cv::parallel_for_(cv::Range(0, K), [&](const cv::Range& r) {
+      for (int i = r.start; i < r.end; i++) {
+        float* data = heatmap.data<float>() + i * H * W;
+        cv::Mat work = cv::Mat(H, W, CV_32FC(1), data);
+        cv::GaussianBlur(work, work, {kernel, kernel}, 0);  // inplace
+      }
+    });
+    std::for_each_n(heatmap.data<float>(), K * H * W, [](float& x) {
```
Comment (lvhan028) [inline csrc/codebase/mmpose/keypoints_from_heatmap.cpp:225]: need to #include <algorithm>

File: csrc/codebase/mmpose/keypoints_from_heatmap.cpp [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+
+    // Transform back to the image
+    for (int i = 0; i < K; i++) {
+      transform_pred(pred, i, center, scale, {W, H}, use_udp);
+    }
+
+    // for (int i = 0; i < K; i++) {
+    //   float *data = pred.data<float>() + i * 3;
+    //   std::cout << *(data + 0) << " " << *(data + 1) << " " << *(data + 2) << "\n";
+    // }
```
Comment (AllentDan) [inline csrc/codebase/mmpose/keypoints_from_heatmap.cpp:202]: may get removed

File: csrc/codebase/mmpose/keypoints_from_regression.cpp [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+namespace mmdeploy::mmpose {
+
+using std::string;
+using std::vector;
+
+class DeepposeRegressionHeadDecode : public MMPose {
+ public:
+  explicit DeepposeRegressionHeadDecode(const Value& config) : MMPose(config) {
+    if (config.contains("params")) {
+      // No params
```
Comment (AllentDan) [inline csrc/codebase/mmpose/keypoints_from_regression.cpp:26]: why an empty condition?

File: csrc/codebase/mmpose/keypoints_from_regression.cpp [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+    // Transform back to the image
+    for (int i = 0; i < K; i++) {
+      transform_pred(pred, i, center, scale, img_size, false);
+    }
+
+    // for (int i = 0; i < K; i++) {
+    //   float* data = pred.data<float>() + i * 3;
+    //   std::cout << *(data + 0) << " " << *(data + 1) << " " << *(data + 2) << "\n";
+    // }
```
Comment (AllentDan) [inline csrc/codebase/mmpose/keypoints_from_regression.cpp:96]: may get removed

File: csrc/codebase/mmpose/keypoints_from_heatmap.cpp [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+        cv::parallel_for_(cv::Range(0, K), [&](const cv::Range& r) {
+          for (int i = r.start; i < r.end; i++) {
+            int kt = (i % 3 == 0) ? 2 * modulate_kernel + 1 : modulate_kernel;
+            float* data = heatmap.data<float>() + i * H * W;
+            cv::Mat work = cv::Mat(H, W, CV_32FC(1), data);
+            cv::GaussianBlur(work, work, {kt, kt}, 0);  // inplace
+          }
+        });
+        float valid_radius = valid_radius_factor_ * H;
+        TensorDesc desc = {Device{"cpu"}, {DataType::kFLOAT}, {1, K / 3, H, W}};
```
Comment (grimoire) [inline csrc/codebase/mmpose/keypoints_from_heatmap.cpp:121]: compile failed with error log:
```
error: braces around scalar initializer for type ‘mmdeploy::DataType’
         TensorDesc desc = {Device{"cpu"}, {DataType::kFLOAT}, {1, K / 3, H, W}};
```

File: csrc/codebase/mmpose/keypoints_from_heatmap.cpp [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+        Tensor offset_x(desc);
+        Tensor offset_y(desc);
+        Tensor heatmap_(desc);
+        {
+          // split heatmap
+          float* src = heatmap.data<float>();
+          float* dst0 = heatmap_.data<float>();
+          float* dst1 = offset_x.data<float>();
+          float* dst2 = offset_y.data<float>();
+          for (int i = 0; i < K; i++) {
```
Comment (lvhan028) [inline csrc/codebase/mmpose/keypoints_from_heatmap.cpp:131]: I think we can simplify this loop, removing if-elseif-else branch
```C++
for (i = 0; i < K / 3; ++i) {
    // 1. [src, src+H*W) to dst0;
    // 2. [src+H*W, src+2*H*W) to dst1;
    // 3. [src+2*H*W, src+3*H*W) to dst2;
}
```

Comment (irexyc) [reply csrc/codebase/mmpose/keypoints_from_heatmap.cpp]: done

File: mmdeploy/codebase/mmpose/deploy/pose_detection_model.py [hunk 8]
```diff
@@ -136,6 +150,74 @@ def show_result(self,
...
+        """
+        scale = _scale.copy()
+        scale = scale / padding * 200
+        center = _center.copy()
+        # fake box
+        box = np.array([center - 0.5 * scale,
+                        center + 0.5 * scale - 1]).flatten()
+        return box
+
+    def forward(self, img: List[torch.Tensor], *args, **kwargs) -> list:
```
Comment (grimoire) [inline mmdeploy/codebase/mmpose/deploy/pose_detection_model.py:182]: Please add more detail docstring about why do we need a new forward

**Review by lvhan028 — APPROVED** [review]
LGTM

File: mmdeploy/codebase/mmpose/deploy/pose_detection_model.py [hunk 9]
```diff
@@ -4,11 +4,22 @@
...
+from mmdeploy.utils import (Backend, get_backend, get_codebase_config,
+                            load_config)
 
 
+def __build_backend_model(cls_name: str, registry: Registry, *args, **kwargs):
+    return registry.module_dict[cls_name](*args, **kwargs)
+
+
+__BACKEND_MODEL = mmcv.utils.Registry(
+    'backend_text_recognizer', build_func=__build_backend_model)
```
Comment (lvhan028) [inline mmdeploy/codebase/mmpose/deploy/pose_detection_model.py:19]: 'backend_text_recognizer' is inappropriate.

File: csrc/codebase/mmpose/keypoints_from_heatmap.cpp (hunk 3 above)
Comment (irexyc) [reply csrc/codebase/mmpose/keypoints_from_heatmap.cpp]: has removed

File: mmdeploy/codebase/mmpose/deploy/pose_detection_model.py (hunk 8 above)
Comment (irexyc) [reply mmdeploy/codebase/mmpose/deploy/pose_detection_model.py:188]: done

File: csrc/codebase/mmpose/topdown_affine.cpp [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+    Device host{"cpu"};
+    auto _img = input["img"].get<Tensor>();
+    OUTCOME_TRY(auto img, MakeAvailableOnDevice(_img, host, stream_));
+    stream_.Wait().value();
+    auto src = cpu::Tensor2CVMat(img);
+
+    // prepare data
+    vector<float> box;
+    from_value(input["box"], box);
+    vector<float> c;
```
[the remark this answers is not in the document]
Comment (irexyc) [reply csrc/codebase/mmpose/topdown_affine.cpp]: add comments

File: csrc/codebase/mmpose/keypoints_from_heatmap.cpp (hunk 6 above)
Comment (irexyc) [reply csrc/codebase/mmpose/keypoints_from_heatmap.cpp]: done

File: csrc/codebase/mmpose/keypoints_from_regression.cpp (hunk 4 above)
Comment (irexyc) [reply csrc/codebase/mmpose/keypoints_from_regression.cpp]: removed

File: csrc/codebase/mmpose/keypoints_from_regression.cpp (hunk 5 above)
Comment (irexyc) [reply csrc/codebase/mmpose/keypoints_from_regression.cpp]: removed

File: csrc/codebase/mmpose/keypoints_from_heatmap.cpp (hunk 2 above)
Comment (irexyc) [reply csrc/codebase/mmpose/keypoints_from_heatmap.cpp]: changed to gcc7.5

File: csrc/apis/c/CMakeLists.txt (hunk 1 above)
Comment (irexyc) [reply csrc/apis/c/CMakeLists.txt]: changed to pose_detector

File: mmdeploy/codebase/mmpose/deploy/pose_detection_model.py (hunk 9 above)
Comment (irexyc) [reply mmdeploy/codebase/mmpose/deploy/pose_detection_model.py]: changed to backend_pose_detectors

**Review by lzhangzz — DISMISSED** [review]
looks ok to me

**Review by grimoire — APPROVED** [review]
LGTM

File: csrc/apis/c/pose_detector.h [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+ * mmdeploy_pose_detector_create_by_path
+ * @param[in] images a batch of images
+ * @param[in] count number of images in the batch
+ * @param[out] results a linear buffer contains the pose result, must be release
+ * by \ref mmdeploy_pose_detector_release_result
+ * @return status code of the operation
+ */
+MMDEPLOY_API int mmdeploy_pose_detector_apply(mm_handle_t handle, const mm_mat_t* mats,
+                                              int mat_count, mm_pose_estimate_t** results,
+                                              int** result_count);
```
[the remark this answers is not in the document]
Comment (irexyc) [reply csrc/apis/c/pose_detector.h]: not used and removed.

**Review by lzhangzz — APPROVED** [review]

**Review by lvhan028 — APPROVED** [review]

**Pull request merged**